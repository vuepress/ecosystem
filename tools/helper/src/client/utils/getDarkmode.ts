/**
 * Get darkmode status from DOM
 *
 * @returns darkmode status
 */
export const getDarkmode = (): boolean =>
  document.documentElement.getAttribute('data-theme') === 'dark'
