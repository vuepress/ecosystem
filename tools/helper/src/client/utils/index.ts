export * from './data.js'
export * from './hasGlobalComponent.js'
export * from './wait.js'
export * from './getHeaders.js'
