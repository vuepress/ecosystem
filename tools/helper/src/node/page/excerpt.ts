/* eslint-disable @typescript-eslint/no-unsafe-enum-comparison */
import { isHTMLTag, isMathMLTag, isSVGTag } from '@vue/shared'
import { load } from 'cheerio'
import type { AnyNode, Element } from 'domhandler'
import matter from 'gray-matter'
import type { App, Page } from 'vuepress/core'
import { isLinkHttp, removeEndingSlash } from 'vuepress/shared'
import { isArray, isLinkAbsolute, startsWith } from '../../shared/index.js'

const HEADING_TAGS = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']

export interface PageExcerptOptions {
  /**
   * Excerpt separator
   *
   * 摘要分隔符
   *
   * @default "<!-- more -->"
   */
  separator?: string

  /**
   * Length of excerpt
   *
   * @description Excerpt length will be the minimal possible length reaching this value
   *
   * 摘要的长度
   *
   * @description 摘要的长度会尽可能的接近这个值
   *
   * @default 300
   */
  length?: number

  /**
   * Tags which is considered as custom elements
   *
   * @description This is used to determine whether a tag is a custom element since all unknown tags are removed in excerpt.
   *
   * 被认为是自定义元素的标签
   *
   * @description 用于判断一个标签是否是自定义元素，因为在摘要中，所有的未知标签都会被移除。
   */
  isCustomElement?: (tagName: string) => boolean

  /**
   * Whether keep page title (first h1) in excerpt
   *
   * 是否保留页面标题 (第一个 h1)
   *
   * @default false
   */
  keepPageTitle?: boolean

  /**
   * Whether preserve tags like line numbers and highlight lines for code blocks
   *
   * 是否保留代码块的标签，诸如行号和高亮行
   *
   * @default false
   */
  keepFenceDom?: boolean
}

interface NodeOptions
  extends Required<
    Pick<PageExcerptOptions, 'isCustomElement' | 'keepFenceDom'>
  > {
  base: string
}

const handleNode = (
  node: AnyNode,
  { base, isCustomElement, keepFenceDom }: NodeOptions,
): AnyNode | null => {
  if (node.type === 'tag') {
    // image using relative urls shall be dropped
    if (node.tagName === 'img') {
      const { src } = node.attribs

      // this is not a resolvable image link
      if (!isLinkHttp(src) && !isLinkAbsolute(src)) return null
    }

    // toc should be dropped
    if (
      [node.attribs.class, node.attribs.id].some((item) =>
        ['table-of-contents', 'toc'].includes(item),
      )
    )
      return null

    // standard tags can be returned
    if (
      isHTMLTag(node.tagName) ||
      isSVGTag(node.tagName) ||
      isMathMLTag(node.tagName)
    ) {
      // handing heading tags
      if (HEADING_TAGS.includes(node.tagName)) {
        // remove heading id tabindex
        delete node.attribs.id
        delete node.attribs.tabindex

        // extract heading tags and remove anchor
        if (
          node.children.length === 1 &&
          node.children[0].type === 'tag' &&
          node.children[0].tagName === 'a' &&
          node.children[0].attribs.class === 'header-anchor'
        )
          node.children = (node.children[0].children[0] as Element).children
      }

      if (
        node.tagName === 'div' &&
        startsWith(node.attribs.class, 'language-')
      ) {
        const pre = node.children.find(
          (childNode) =>
            childNode.type === 'tag' &&
            childNode.tagName === 'pre' &&
            startsWith(childNode.attribs.class, 'language-'),
        )

        if (
          // we are sure this is a code fence
          pre &&
          !keepFenceDom
        ) {
          node.attribs.class = node.attribs.class.replace(
            ' line-numbers-mode',
            '',
          )
          node.children = [pre]
        }
      }

      // remove `v-pre` attribute on code
      if (node.tagName === 'code' || node.tagName === 'pre')
        delete node.attribs['v-pre']

      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      node.children = handleNodes(node.children, {
        base,
        isCustomElement,
        keepFenceDom,
      })

      return node
    }

    // we shall convert `<RouteLink>` and `<RouterLink>` to `<a>` tag
    if (node.tagName === 'routelink' || node.tagName === 'routerlink') {
      node.tagName = 'a'
      node.attribs.href = `${removeEndingSlash(base)}${node.attribs.to}`
      node.attribs.target = '_blank'
      delete node.attribs.to

      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      node.children = handleNodes(node.children, {
        base,
        isCustomElement,
        keepFenceDom,
      })

      return node
    }

    // keep custom element as is
    if (isCustomElement(node.tagName)) return node

    // other tags are considered as vue components and dropped
    return null
  }

  return node
}

const handleNodes = (
  nodes: AnyNode[] | null,
  { base, isCustomElement, keepFenceDom }: NodeOptions,
): AnyNode[] =>
  isArray(nodes)
    ? nodes
        .map((node) =>
          handleNode(node, {
            base,
            isCustomElement,
            keepFenceDom,
          }),
        )
        .filter((node): node is AnyNode => node !== null)
    : []

const $ = load('')

const isH1Tag = (node: AnyNode): boolean =>
  node.type === 'tag' && node.tagName === 'h1'

export const getPageExcerptContent = (
  content: string,
  separator = '<!-- more -->',
): string | undefined =>
  matter(content, {
    excerpt: true,
    excerpt_separator: separator,
  }).excerpt

export const getPageExcerpt = (
  { markdown, options: { base } }: App,
  { content, contentRendered, filePath, filePathRelative, frontmatter }: Page,
  {
    isCustomElement = (): boolean => false,
    separator = '<!-- more -->',
    length = 300,
    keepPageTitle = false,
    keepFenceDom = false,
  }: PageExcerptOptions = {},
): string => {
  // get page content
  const excerpt = getPageExcerptContent(content, separator)

  if (excerpt) {
    const renderedContent = markdown.render(
      excerpt,
      // markdown env
      {
        base,
        filePath,
        filePathRelative,
        frontmatter: { ...frontmatter },
      },
    )

    const rootNodes = renderedContent ? $.parseHTML(renderedContent) : []

    if (rootNodes[0] && !keepPageTitle && isH1Tag(rootNodes[0]))
      rootNodes.shift()

    return $.html(
      handleNodes(rootNodes, {
        base,
        isCustomElement,
        keepFenceDom,
      }),
    )
  }

  if (length > 0) {
    let autoExcerpt = ''

    const rootNodes = contentRendered ? $.parseHTML(contentRendered) : []

    if (rootNodes[0] && !keepPageTitle && isH1Tag(rootNodes[0]))
      rootNodes.shift()

    for (const node of rootNodes) {
      const resolvedNode = handleNode(node, {
        base,
        isCustomElement,
        keepFenceDom,
      })

      if (resolvedNode) {
        autoExcerpt += $.html(resolvedNode)
        if (autoExcerpt.length >= length) break
      }
    }

    return autoExcerpt
  }

  return ''
}
