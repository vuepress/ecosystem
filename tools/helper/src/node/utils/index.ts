export * from './data.js'
export * from './isModuleAvailable.js'
export * from './getModulePath.js'
export * from './logger.js'
export * from './packageManager.js'
