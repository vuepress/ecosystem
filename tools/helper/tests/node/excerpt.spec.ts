import { describe, expect, it } from 'vitest'
import { createBaseApp } from 'vuepress/core'
import { path } from 'vuepress/utils'
import { getPageExcerpt } from '../../src/node/page/excerpt.js'
import type { PageExcerptOptions } from '../../src/node/page/excerpt.js'
import { emptyTheme } from '../__fixtures__/theme/empty.js'

describe('getPageExcerpt', async () => {
  const app = createBaseApp({
    bundler: {} as any,
    source: path.resolve(__dirname, '../__fixtures__/src'),
    theme: emptyTheme,
  })

  await app.init()

  const getExcerptData = (options: PageExcerptOptions = {}) =>
    app.pages
      .filter(({ path }) => path !== '/404.html')
      .map((page) => ({
        path: page.path,
        excerpt: getPageExcerpt(app, page, options),
      }))

  describe('default', () => {
    const excerptData = getExcerptData()

    it('generate excerpt for all pages', () => {
      excerptData.forEach(({ excerpt, path }) => {
        expect(excerpt.length).toBeGreaterThan(0)
        expect(excerpt).toMatchSnapshot(path)
      })
    })

    describe('handle contents correctly', () => {
      getExcerptData({ length: Infinity }).forEach(({ excerpt, path }) => {
        if (path === '/markdown.html') {
          it('clean heading', () => {
            ;[1, 2, 3, 4, 5, 6].forEach((i) => {
              expect(excerpt).toContain(`<h${i}>Heading ${i}</h${i}>`)
            })

            expect(excerpt).toContain(
              '<h3>Heading with <strong>markdown</strong> and <span>html</span></h3>',
            )
          })

          it('handle inline tags', () => {
            expect(excerpt).toContain('<strong>bold</strong>')
            expect(excerpt).toContain('<em>italic</em>')
            expect(excerpt).toContain('<s>delete</s>')
          })

          it('handle block tags', () => {
            expect(excerpt).toContain('<p>')
            expect(excerpt).toContain('<ul>')
            expect(excerpt).toContain('<ol>')
            expect(excerpt).toContain('<li>')
            expect(excerpt).toContain('<blockquote>')
          })

          it('handle links', () => {
            expect(excerpt).toContain(
              '<a href="/excerpt.html" target="_blank">relative markdown link</a>',
            )
            expect(excerpt).toContain(
              '<a href="/excerpt.html" target="_blank">relative html link</a>',
            )
            expect(excerpt).toContain(
              '<a href="/" target="_blank">absolute markdown link</a>',
            )
            expect(excerpt).toContain(
              '<a href="/" target="_blank">absolute html link</a>',
            )
            expect(excerpt).toContain('<a href="#link">Anchor</a>')
          })

          it('remove relative image', () => {
            expect(excerpt).not.toContain('relative.jpg')
          })

          it('preserve image with url or absolute links', () => {
            expect(excerpt).toContain('src="https://exmaple.com/logo.png"')
            expect(excerpt).toContain('src="/logo.png"')
          })

          it('inline code', () => {
            expect(excerpt).toContain('<code>inline code</code>')
          })

          it('code fence', () => {
            expect(excerpt).toContain('<code>inline code</code>')
          })

          it('remove comment', () => {
            expect(excerpt).toContain('Sample text here...')
          })
        }
      })
    })

    it('remove unknown tags', () => {
      excerptData.forEach(({ excerpt, path }) => {
        if (path === '/component.html') {
          expect(excerpt).not.toContain('custom-element')
          expect(excerpt).not.toContain('VueComponent')
        }
      })
    })
  })

  describe('excerptLength', () => {
    it('only generate when having marker with 0', () => {
      getExcerptData({ length: 0 }).forEach(({ excerpt, path }) => {
        if (path === '/separator.html') {
          expect(excerpt).toContain('article excerpt')
          expect(excerpt).not.toContain('main content')
          expect(excerpt).toMatchSnapshot(path)
        } else {
          expect(excerpt.length).toBe(0)
        }
      })
    })

    it('extract all content with Infinity', () => {
      getExcerptData({ length: Infinity }).forEach(({ excerpt, path }) => {
        expect(excerpt.length).toBeGreaterThan(0)
        expect(excerpt).toMatchSnapshot(path)

        if (path === '/long-content.html') {
          expect(excerpt).toContain('Content ends.')
        }
      })
    })
  })

  describe('excerptSeparator', () => {
    it('generate excerpt with default marker', () => {
      getExcerptData().forEach(({ excerpt, path }) => {
        if (path === '/separator.html') {
          expect(excerpt).toContain('article excerpt')
          expect(excerpt).not.toContain('main content')
          expect(excerpt).toMatchSnapshot(path)
        }
      })
    })

    it('generate excerpt with custom marker', () => {
      getExcerptData({ separator: 'END_OF_EXCERPT' }).forEach(
        ({ excerpt, path }) => {
          if (path === '/custom-separator.html') {
            expect(excerpt).toContain('article excerpt')
            expect(excerpt).not.toContain('main content')
            expect(excerpt).toMatchSnapshot(path)
          }
        },
      )
    })
  })

  describe('isCustomElement', () => {
    it('preserve custom element', () => {
      getExcerptData({
        isCustomElement: (tag) => tag.startsWith('custom-element'),
      }).forEach(({ excerpt, path }) => {
        if (path === '/component.html') {
          expect(excerpt).toContain('custom-element1')
          expect(excerpt).toContain('custom-element2')
          expect(excerpt).toContain('custom-element3')
          expect(excerpt).not.toContain('VueComponent')
        }
      })
    })
  })

  describe('keepPageTitle', () => {
    it('remove first h1', () => {
      getExcerptData().forEach(({ excerpt, path }) => {
        if (path === '/markdown.html') {
          expect(excerpt).not.toContain('Content Example')
        }
      })
    })

    it('not remove first h1', () => {
      getExcerptData({ keepPageTitle: true }).forEach(({ excerpt, path }) => {
        if (path === '/markdown.html') {
          expect(excerpt).toContain('Content Example')
        }
      })
    })
  })
})
