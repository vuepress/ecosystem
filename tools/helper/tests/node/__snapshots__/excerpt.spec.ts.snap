// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPageExcerpt > default > generate excerpt for all pages > / 1`] = `
"<p>Homepage Content</p>
"
`;

exports[`getPageExcerpt > default > generate excerpt for all pages > /component.html 1`] = `
"<p>A text with </p>
"
`;

exports[`getPageExcerpt > default > generate excerpt for all pages > /custom-separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
<p>END_OF_EXCERPT</p>
<h2>Content</h2>
<p>Here is main content of <strong>article</strong>.</p>
<ol>
<li>A</li>
<li>B</li>
<li>C</li>
</ol>
<pre><code class="language-js">const a = 1
</code></pre>"
`;

exports[`getPageExcerpt > default > generate excerpt for all pages > /long-content.html 1`] = `
"<p>Lorem ipsum dolor sit amet consectetur adipisicing elit. Nemo, rem. Recusandae itaque rem, non perspiciatis magnam molestiae, in ab quae, dolore illo neque possimus nisi inventore? Laudantium officia nihil iusto!</p>
<p>Quod delectus vero dicta perferendis quia sint incidunt vitae pariatur, nemo suscipit saepe veritatis laudantium deleniti aperiam! Explicabo dignissimos, corporis, odio voluptatum eius aut pariatur dicta inventore, amet optio modi!</p>"
`;

exports[`getPageExcerpt > default > generate excerpt for all pages > /markdown.html 1`] = `
"
<h2>Headings</h2>
<h1>Heading 1</h1>
<h2>Heading 2</h2>
<h3>Heading 3</h3>
<h4>Heading 4</h4>
<h5>Heading 5</h5>
<h6>Heading 6</h6>
<h3>Heading with <strong>markdown</strong> and <span>html</span></h3>
<h2>Text</h2>
<p>This sentence has <strong>bold</strong>„ÄÅ<em>italic</em> and <s>delete</s> style text.</p>"
`;

exports[`getPageExcerpt > default > generate excerpt for all pages > /separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`getPageExcerpt > excerptLength > extract all content with Infinity > / 1`] = `
"<p>Homepage Content</p>
"
`;

exports[`getPageExcerpt > excerptLength > extract all content with Infinity > /component.html 1`] = `
"<p>A text with </p>
"
`;

exports[`getPageExcerpt > excerptLength > extract all content with Infinity > /custom-separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
<p>END_OF_EXCERPT</p>
<h2>Content</h2>
<p>Here is main content of <strong>article</strong>.</p>
<ol>
<li>A</li>
<li>B</li>
<li>C</li>
</ol>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`getPageExcerpt > excerptLength > extract all content with Infinity > /long-content.html 1`] = `
"<p>Lorem ipsum dolor sit amet consectetur adipisicing elit. Nemo, rem. Recusandae itaque rem, non perspiciatis magnam molestiae, in ab quae, dolore illo neque possimus nisi inventore? Laudantium officia nihil iusto!</p>
<p>Quod delectus vero dicta perferendis quia sint incidunt vitae pariatur, nemo suscipit saepe veritatis laudantium deleniti aperiam! Explicabo dignissimos, corporis, odio voluptatum eius aut pariatur dicta inventore, amet optio modi!</p>
<p>Eligendi voluptatum animi aspernatur rerum hic quasi neque doloribus mollitia quaerat. Id blanditiis, nostrum autem sequi vero praesentium magnam totam in sit animi velit, veritatis hic natus? Quidem, distinctio provident!</p>
<p>Nemo distinctio numquam reprehenderit sed alias perspiciatis autem quasi qui culpa ducimus dolorum mollitia beatae aperiam quisquam necessitatibus neque voluptatum sapiente, quidem explicabo voluptate debitis laborum delectus. Hic, ipsam non?</p>
<p>Vitae saepe repellat, laboriosam ipsam numquam cumque, eaque perferendis quibusdam dolorem blanditiis, architecto velit molestiae. Earum ea cupiditate excepturi incidunt sequi ducimus ullam architecto, tempora, iure molestias in repellendus hic.</p>
<p>Vero ducimus minima vitae voluptas in magnam fuga ipsa a debitis alias porro non earum veniam laudantium ipsam perferendis itaque, voluptatum amet provident. Deleniti voluptatem sit numquam ducimus quam cupiditate!</p>
<p>Molestias cum, rerum perferendis at voluptatibus est assumenda. Soluta saepe corrupti sunt fugit laboriosam possimus labore nemo nisi veritatis neque minima qui, molestiae esse commodi sit iure voluptatem voluptatum quibusdam.</p>
<p>Asperiores fuga magnam explicabo sed quae laboriosam ullam aut harum laborum accusantium totam reprehenderit voluptatibus, vel facere, inventore soluta consectetur sint? Maxime nobis dicta a quos, voluptate quidem odit quo?</p>
<p>Eos magni ullam aperiam provident ab repellat sunt accusantium recusandae quisquam modi quasi quaerat et, officiis quis dolorem optio minima? Dicta aut dolorem eius, aliquid maiores labore! Officiis, tempore similique.</p>
<p>Ex quo in nulla molestias doloribus odio dolorem illum aperiam error iste! Facilis quae consequatur praesentium unde. Rerum, et, doloribus eveniet cum sequi autem illo est quibusdam laborum aperiam nostrum.</p>
<p>Eius ratione fuga dolorum aut totam molestias commodi ea sed odio possimus, fugiat perspiciatis perferendis omnis autem deserunt. Id veniam rem voluptatum non molestiae adipisci, impedit cupiditate eos exercitationem nisi.</p>
<p>Ab nam veniam corrupti aperiam cupiditate. Ex totam, cumque dignissimos dicta nemo veritatis facilis dolor minus consequatur et quo odio quibusdam consequuntur dolorum esse consectetur dolorem velit, perferendis deserunt perspiciatis!</p>
<p>Repellendus in labore sed nobis esse aut nam, neque incidunt asperiores natus fuga, reprehenderit porro ad, atque nemo dicta autem impedit iste totam quae minus? Ullam nesciunt iste voluptatum deserunt?</p>
<p>Corporis numquam itaque quibusdam eos quas tempore dolorem, impedit laboriosam sequi, possimus sint, voluptatum ab dolor unde! Harum illo, sed, adipisci iusto saepe provident, pariatur labore obcaecati dolorem velit nostrum.</p>
<p>Rerum minima eius hic, vel iure ipsa natus facilis necessitatibus dolores optio enim quod obcaecati neque doloribus sapiente atque repellendus. Amet eligendi libero illo? Earum aliquam molestiae tempora placeat delectus!</p>
<p>Dolores eum hic accusantium aliquam maxime deleniti non, voluptas aut cupiditate omnis veritatis dolorem, voluptatem nisi vero odio quidem sapiente animi! Nihil repudiandae, excepturi totam cupiditate blanditiis porro delectus quidem.</p>
<p>Iste ipsum qui, nostrum recusandae, maiores, at cum odio labore cupiditate numquam id animi iure? Quam sequi laborum vero magnam excepturi, mollitia ex nobis fuga sed, vel veniam molestiae hic?</p>
<p>Iste temporibus voluptate magni voluptates inventore deserunt magnam laboriosam rerum exercitationem maxime, minus sed accusamus sint perspiciatis quaerat, ipsa facilis ipsum? Commodi incidunt nam doloremque blanditiis neque minima beatae aliquam.</p>
<p>Veniam voluptate exercitationem voluptas blanditiis aliquam id non fugiat amet, esse iusto dolorem deserunt maxime libero, similique ducimus, reiciendis adipisci ad culpa! Unde corporis, magnam quo iusto blanditiis ex voluptatibus.</p>
<p>Ipsum quam accusamus atque itaque reiciendis obcaecati nemo quas architecto, suscipit, dolores a quasi illo at deserunt voluptas nesciunt. Odio sapiente dolores beatae et labore dicta aspernatur molestiae velit rerum?</p>
<p>Laborum vel vitae dignissimos enim amet quas eveniet aliquid earum numquam itaque possimus, est corrupti quidem quam natus nostrum neque repellat quis ratione corporis officiis minus expedita dicta voluptate! Quibusdam!</p>
<p>Qui minima provident dignissimos corrupti facilis modi saepe esse praesentium itaque molestiae aut, ea, earum, iure facere quis eaque alias! Voluptas in minima voluptate quos facilis perspiciatis expedita accusamus sequi!</p>
<p>Amet, sed qui eius delectus blanditiis possimus fugit et nulla doloribus, tempore beatae autem tenetur deserunt eveniet vero dolore sit ipsa minus magnam itaque aliquid labore, illo reprehenderit dolores! Sunt.</p>
<p>Commodi aliquam consectetur saepe quis laboriosam, debitis optio tempora eius sed error ad officiis. Cupiditate enim amet laborum quaerat eaque. Fugit ab ipsam similique tempore nesciunt accusantium explicabo repudiandae harum?</p>
<p>Autem atque laudantium mollitia voluptates eum corporis repudiandae debitis dicta cumque asperiores repellat accusamus sit esse, dolor, inventore libero minima tempora delectus eligendi quas sint! Odit eos ipsa accusamus ducimus.</p>
<p>Dicta natus facere, cum eveniet explicabo nobis, pariatur beatae quae ab temporibus maiores rem eaque. Quisquam itaque saepe sequi, voluptatem quibusdam, minima ea deleniti quos vel modi est. Modi, maxime?</p>
<p>Optio perspiciatis rerum iusto expedita esse magnam pariatur quo repellendus sint tenetur veniam illo alias quae ratione aliquid ab natus enim laboriosam veritatis qui possimus, quos dolorum officia quas? Quasi.</p>
<p>Consectetur, incidunt praesentium laborum consequatur magnam, accusantium reprehenderit sequi explicabo nam numquam iusto veniam ducimus temporibus molestiae ab nisi quis dolores quam unde repellat sunt, beatae quidem? Doloremque, vero minima?</p>
<p>Illo reiciendis veritatis rem adipisci dolores quasi dignissimos quisquam, eligendi sint facilis est perspiciatis velit totam in praesentium ipsum! Accusamus dignissimos minima fuga laudantium pariatur voluptate dolores veniam deserunt beatae?</p>
<p>Id, molestiae. Excepturi ad quaerat quam sit consequuntur dolores, a doloribus eaque et harum nulla illo alias beatae, temporibus non maiores quo! Numquam aliquid, enim quas possimus ab exercitationem accusamus.</p>
<p>Ab voluptates consequatur dolore aperiam, eum doloremque voluptatum dicta ipsum eos ea exercitationem. Soluta nobis sunt reiciendis consequuntur vero nesciunt suscipit quos maiores saepe, commodi doloremque, culpa rerum perferendis esse.</p>
<p>Quasi saepe sed repellendus. Fugiat voluptatem aspernatur nobis? Officiis necessitatibus veniam esse provident voluptatum alias quas consequuntur corporis aliquam doloribus eveniet autem, sit, quo dolorum voluptate vel porro quae quibusdam.</p>
<p>Aliquid illo maiores vel cumque eos iure explicabo quae molestiae exercitationem dolores quia voluptate, similique consequatur necessitatibus pariatur quas labore rerum consectetur eum totam dignissimos quaerat. Perferendis dicta facere saepe.</p>
<p>Nam molestias vel natus eaque, sunt itaque ratione maxime libero fuga fugit minus quidem velit assumenda, rem expedita impedit a in! Cumque perferendis est velit quo ullam fugit corrupti pariatur!</p>
<p>Eaque at cumque dolores culpa autem officiis architecto voluptas qui mollitia, nam repellat quis fuga consequuntur? Dolore optio nesciunt error perspiciatis. Praesentium nisi, fugiat adipisci suscipit reprehenderit ullam quia deleniti.</p>
<p>Iusto illo in voluptatum, reprehenderit atque hic inventore asperiores facere enim minima mollitia praesentium assumenda incidunt obcaecati. Delectus neque soluta, ullam nemo corrupti dicta sint explicabo nesciunt tempora, unde pariatur.</p>
<p>Suscipit, placeat, cum veniam recusandae ab dolor corporis laboriosam voluptate consequatur nihil deleniti quae soluta assumenda quam cumque aliquam, possimus expedita explicabo eligendi consequuntur accusantium enim! Maxime rem est ex!</p>
<p>Error ipsam dolorem temporibus, esse suscipit eveniet recusandae hic rem corporis excepturi laboriosam est nostrum? Illum distinctio laboriosam et, itaque temporibus maxime culpa aliquid iure ipsam? Commodi, aliquid quasi. Quod!</p>
<p>Ullam ratione tempora molestias unde, ut, dolor sint tenetur magni amet sunt, nihil corporis voluptatum fuga sapiente soluta maxime voluptas fugiat recusandae assumenda nobis quam maiores veniam cupiditate. Officia, sapiente.</p>
<p>Beatae alias repellendus nostrum sequi porro. Fugit cupiditate aperiam quos ratione dicta eligendi deserunt quisquam iste minima, unde natus dolorum vero. Consequatur similique quasi tenetur. Nisi illo vitae aperiam ipsam?</p>
<p>Praesentium, quidem accusamus doloribus exercitationem adipisci rem soluta asperiores quis corrupti pariatur? Incidunt architecto tempora similique repellat ratione voluptatibus magnam, odit harum maxime exercitationem reprehenderit repudiandae in facilis obcaecati velit.</p>
<p>Minus, rerum, inventore dolorem illo dolor tenetur molestiae sit quae sapiente officia, maxime voluptatum consequuntur ullam. Temporibus voluptates ipsa facere, optio odit saepe nihil repellendus ad reprehenderit expedita consequuntur nam!</p>
<p>Aut corrupti eligendi rem repudiandae cum similique officia ea illum beatae. Maiores, explicabo, voluptatum accusamus excepturi incidunt velit dolorem eligendi distinctio a nemo similique perspiciatis blanditiis dolores voluptates commodi porro?</p>
<p>Quis nulla minima corporis eligendi earum voluptas ipsam accusantium. Nesciunt corrupti numquam ex omnis fuga natus sint, delectus molestias eaque dolor ea porro autem velit, voluptate pariatur exercitationem modi explicabo.</p>
<p>Corporis vero quisquam id mollitia deleniti minus aut libero sint iure sapiente magni nemo laboriosam, voluptatem totam optio nostrum labore tenetur, a nesciunt! Quis impedit harum velit, minima quae maiores.</p>
<p>Cum sint, rem saepe quam similique sequi veniam nisi labore suscipit dolorem quos impedit alias dolore sit ad incidunt assumenda. Blanditiis sed perspiciatis reiciendis animi ea sunt, at perferendis iusto!</p>
<p>Qui voluptas quidem ipsa dolores exercitationem doloremque necessitatibus? Rem nihil deleniti eos, nobis doloremque a voluptatum dignissimos commodi cumque tempora facilis! Illo minima unde beatae dicta nobis ipsa, delectus dolore!</p>
<p>Illum quis laboriosam neque odio ad! Facere illo exercitationem quod in vel non magnam incidunt quibusdam culpa, minus pariatur id eaque dolorum, consequatur voluptas harum voluptatibus sunt odit iure atque.</p>
<p>Dolorum officiis fugit error laborum. Necessitatibus aliquam reprehenderit aspernatur. Beatae illo est reiciendis. Odit nobis pariatur provident, deleniti expedita magni maxime quaerat dolor omnis enim qui ducimus, fugit consectetur odio?</p>
<p>Beatae sunt, quia doloremque nulla animi aperiam facilis blanditiis, non laborum in delectus a ipsa esse ex odio maxime iure, magni excepturi pariatur vero veritatis. Officiis deserunt laboriosam consequuntur recusandae.</p>
<p>Content ends.</p>
"
`;

exports[`getPageExcerpt > excerptLength > extract all content with Infinity > /markdown.html 1`] = `
"
<h2>Headings</h2>
<h1>Heading 1</h1>
<h2>Heading 2</h2>
<h3>Heading 3</h3>
<h4>Heading 4</h4>
<h5>Heading 5</h5>
<h6>Heading 6</h6>
<h3>Heading with <strong>markdown</strong> and <span>html</span></h3>
<h2>Text</h2>
<p>This sentence has <strong>bold</strong>„ÄÅ<em>italic</em> and <s>delete</s> style text.</p>
<h2>Paragraph</h2>
<p>This is a paragraph.</p>
<p>This is another paragraph.</p>
<h2>Line Break</h2>
<p>I would like to line break at<br>
this point</p>
<p>::: tip</p>
<p>In codes above, two spaces are behind <code>at</code>.</p>
<p>:::</p>
<h2>Blockquotes</h2>
<blockquote>
<p>Blockquotes can also be nested...</p>
<blockquote>
<p>...by using greater-than signs right next to each other...</p>
<blockquote>
<p>...or with spaces between arrows.</p>
</blockquote>
</blockquote>
</blockquote>
<h2>List</h2>
<h3>Unordered List</h3>
<ul>
<li>
<p>Create a list by starting a line with <code>-</code></p>
</li>
<li>
<p>Make sub-lists by indenting 2 spaces:</p>
<ul>
<li>
<p>Marker character change forces new list start:</p>
<ul>
<li>
<p>Ac tristique libero volutpat at</p>
</li>
<li>
<p>Facilisis in pretium nisl aliquet</p>
</li>
<li>
<p>Nulla volutpat aliquam velit<br>
link break</p>
<p>New paragraph</p>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p>It‚Äôs easy!</p>
</li>
</ul>
<h3>Ordered List</h3>
<ol>
<li>Lorem ipsum dolor sit amet</li>
<li>Consectetur adipiscing elit<br>
line break<br>
line break again</li>
<li>Integer molestie lorem at massa</li>
</ol>
<h2>HR</h2>
<hr>
<h2>Link</h2>
<ul>
<li><a href="/excerpt.html" target="_blank">relative markdown link</a></li>
<li><a href="/excerpt.html" target="_blank">relative html link</a></li>
<li><a href="/" target="_blank">absolute markdown link</a></li>
<li><a href="/" target="_blank">absolute html link</a></li>
<li><a href="#link">Anchor</a></li>
</ul>
<h2>Image</h2>
<p><img src="/logo.png" alt="Logo"></p>
<p><img src="https://exmaple.com/logo.png" alt="logo"></p>
<p>A inline image , <img src="/logo.png" alt="image with absolute link"> and a <img src="https://exmaple.com/logo.png" alt="image with full url">.</p>
<h2>Emoji</h2>
<p>Classic:</p>
<p>üòâ üò¢ üòÜ üòã</p>
<h2>Tables</h2>
<table>
<thead>
<tr>
<th style="text-align:center">center</th>
<th style="text-align:right">right</th>
<th style="text-align:left">left</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">For center align use <code>:-:</code></td>
<td style="text-align:right">For right align use <code>-:</code></td>
<td style="text-align:left">For left align use <code>:-</code></td>
</tr>
<tr>
<td style="text-align:center">table text</td>
<td style="text-align:right">aaaaaaaaa</td>
<td style="text-align:left">aaaa</td>
</tr>
<tr>
<td style="text-align:center">c</td>
<td style="text-align:right">aaaa</td>
<td style="text-align:left">a</td>
</tr>
</tbody>
</table>
<h2>Codes</h2>
<p>Inline Code: <code>inline code</code></p>
<p>Block code:</p>
<pre><code>Sample text here...
</code></pre>
<p>Syntax highlighting:</p>
<pre><code class="language-js">var foo = function (bar) {
  return bar++
}

console.log(foo(5))
</code></pre>
<h2>Comment</h2>
<!-- Here is a comment -->
"
`;

exports[`getPageExcerpt > excerptLength > extract all content with Infinity > /separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`getPageExcerpt > excerptLength > only generate when having marker with 0 > /separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`getPageExcerpt > excerptSeparator > generate excerpt with custom marker > /custom-separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;

exports[`getPageExcerpt > excerptSeparator > generate excerpt with default marker > /separator.html 1`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<pre><code class="language-js">const a = 1
</code></pre>
"
`;
